presto-main/src/main/java/com/facebook/presto/operator/ExchangeClientFactory.java

```diff
@@ -76,11 +91,28 @@ public ExchangeClientFactory(
 
         this.scheduler = requireNonNull(scheduler, "scheduler is null");
 
+        this.pageBufferClientCallbackExecutor = newCachedThreadPool(daemonThreadsNamed("page-buffer-client-callback-%s"));
+        this.boundedExecutor = new BoundedExecutor(pageBufferClientCallbackExecutor, pageBufferClientMaxCallbackThreads);
```

Comment:
I would just allocate a fixed pool here (there are only 25 by default), since scaling a thread pool when under load is expensive and can fail.